subsys/net/lib/http_server/src/server.c

```diff
@@ -0,0 +1,538 @@
+				   "not support "
+				   "HTTP/2.\n");
+			    handle_http1_request(ctx, i);
+			} else {
+			    handle_http2_request(ctx, i, valread);
+			}
+		    }
+		}
+	    }
+	}
+    }
+
+    return 0;
+}
+
+void generate_response_headers_frame(unsigned char *response_headers_frame,
+				     int new_stream_id)
+{
+    response_headers_frame[0] = 0x00;
+    response_headers_frame[1] = 0x00;
+    response_headers_frame[2] = 0x02;
+    response_headers_frame[3] = 0x01;
+    response_headers_frame[4] = 0x04;
+    response_headers_frame[5] = 0x00;
+    response_headers_frame[6] = 0x00;
+    response_headers_frame[7] = 0x00;
+    response_headers_frame[8] = new_stream_id & 0xFF;
+    response_headers_frame[9] = 0x88;		      /* HPACK :status: 200 */
+    response_headers_frame[10] = 0x90; /* HPACK accept-encoding: gzip, deflate */
+}
+
+void sendData(int socket_fd, const char *payload, size_t length, uint8_t type,
+	      uint8_t flags, uint32_t streamId)
+{
+    if (9 + length > MAX_FRAME_SIZE) {
+	printf("Payload is too large for the buffer\n");
+	return;
+    }
+
+    uint8_t data_frame[MAX_FRAME_SIZE];
+
+    data_frame[0] = (length >> 16) & 0xFF;
+    data_frame[1] = (length >> 8) & 0xFF;
+    data_frame[2] = length & 0xFF;
+
+    data_frame[3] = type;
+    data_frame[4] = flags;
+
+    data_frame[5] = (streamId >> 24) & 0xFF;
+    data_frame[6] = (streamId >> 16) & 0xFF;
+    data_frame[7] = (streamId >> 8) & 0xFF;
+    data_frame[8] = streamId & 0xFF;
+
+    memcpy(data_frame + 9, payload, length);
+
+    int frame_size = 9 + length;
+    if (sendall(socket_fd, (const char *)data_frame, frame_size) < 0) {
+	perror("ERROR writing to socket");
+	exit(EXIT_FAILURE);
+    }
+}
+
+void print_http2_frames(Http2Frame *frames, unsigned int frame_count)
+{
+    for (unsigned int i = 0; i < frame_count; i++) {
+	printf("Frame %d:\n", i);
+	printf("  Length: %u\n", frames[i].length);
+	printf("  Type: %u\n", frames[i].type);
+	printf("  Flags: %u\n", frames[i].flags);
+	printf("  Stream Identifier: %u\n", frames[i].stream_identifier);
+	printf("  Payload: ");
+	for (unsigned int j = 0; j < frames[i].length; j++) {
+	    printf("%02x ", frames[i].payload[j]);
+	}
+	printf("\n\n");
+    }
+}
+
+unsigned int parse_http2_frames(unsigned char *buffer, unsigned long buffer_len,
+				Http2Frame *frames)
+{
+    unsigned int frame_count = 0;
+    unsigned long pos = 0;
+
+    while (pos + 9 <= buffer_len && frame_count < MAX_FRAMES) {
+	frames[frame_count].length =
+	    (buffer[pos] << 16) | (buffer[pos + 1] << 8) | buffer[pos + 2];
+	frames[frame_count].type = buffer[pos + 3];
+	frames[frame_count].flags = buffer[pos + 4];
+	frames[frame_count].stream_identifier =
+	    (buffer[pos + 5] << 24) | (buffer[pos + 6] << 16) |
+	    (buffer[pos + 7] << 8) | buffer[pos + 8];
+	frames[frame_count].stream_identifier &= 0x7FFFFFFF;
+
+	pos += 9;
+
+	/* Ckeck if payload size doesn't exceed buffer size */
+	if (pos + frames[frame_count].length > buffer_len) {
+	    break;
```

Comment:
We still should be prepared for this case. I. e. the actual paylaod size may exceed the buffer we're processing, so that in case of the next `recv()` call we should treat the remainig data as a payload for the previous frame, and not try to parse the new frame header.